File: solutions/cpp-solutions/test/1-two-sum-test.cpp
#include "leetcode/1-two-sum.h"
#include <gtest/gtest.h>

TEST(_1TwoSumTest, TestCase1) {
    std::vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    std::vector<int> expected = {0, 1};

    auto solution = _1TwoSum::Solution();
    EXPECT_EQ(solution.twoSumBrute(nums, target), expected);
    EXPECT_EQ(solution.twoSumSort(nums, target), expected);
    EXPECT_EQ(solution.twoSumMap(nums, target), expected);
}

TEST(_1TwoSumTest, TestCase2) {
    std::vector<int> nums = {3, 2, 4};
    int target = 6;
    std::vector<int> expected = {1, 2};

    auto solution = _1TwoSum::Solution();
    EXPECT_EQ(solution.twoSumBrute(nums, target), expected);
    EXPECT_EQ(solution.twoSumSort(nums, target), expected);
    EXPECT_EQ(solution.twoSumMap(nums, target), expected);
}

TEST(_1TwoSumTest, TestCase3) {
    std::vector<int> nums = {3, 3};
    int target = 6;
    std::vector<int> expected = {0, 1};

    auto solution = _1TwoSum::Solution();
    EXPECT_EQ(solution.twoSumBrute(nums, target), expected);
    EXPECT_EQ(solution.twoSumSort(nums, target), expected);
    EXPECT_EQ(solution.twoSumMap(nums, target), expected);
}

File: solutions/typescript-solutions/test/solutions/1-two-sum.test.ts
import { describe, it } from 'node:test';
import { strict as assert } from 'node:assert';
import { twoSumBrute, twoSumSort, twoSumHash } from '../../src/problems/1-two-sum.js';

describe('twoSum', () => {
    it('1st test case', () => {
        const nums = [2, 7, 11, 15];
        const target = 9;
        const expected = [0, 1];
        assert.deepEqual(twoSumBrute(nums, target), expected);
        assert.deepEqual(twoSumSort(nums, target), expected);
        assert.deepEqual(twoSumHash(nums, target), expected);
    });

    it('2nd test case', () => {
        const nums = [3, 2, 4];
        const target = 6;
        const expected = [1, 2];
        assert.deepEqual(twoSumBrute(nums, target), expected);
        assert.deepEqual(twoSumSort(nums, target), expected);
        assert.deepEqual(twoSumHash(nums, target), expected);
    });

    it('3nd test case', () => {
        const nums = [3, 3];
        const target = 6;
        const expected = [0, 1];
        assert.deepEqual(twoSumBrute(nums, target), expected);
        assert.deepEqual(twoSumSort(nums, target), expected);
        assert.deepEqual(twoSumHash(nums, target), expected);
    });

    it('throws error when no solutions', () => {
        const nums = [1, 2, 3];
        const target = 6;
        assert.throws(() => twoSumBrute(nums, target));
        assert.throws(() => twoSumSort(nums, target));
        assert.throws(() => twoSumHash(nums, target));
    });
});
